packages/editor/src/optimizer/common.ts

```diff
@@ -0,0 +1,614 @@
+  | CompileResult
+  | PollResult
+  | ComputeLayoutResult
+  | ResampleResult;
+
+/** Data contained in a `Notification` */
+export type NotificationData =
+  | InitData
+  | LabelMeasurementData
+  | DiscardDiagramData;
+
+/** Ensure request data and result pairs have the same tag, to check validity */
+export enum MessageTags {
+  Compile = "Compile",
+  Poll = "Poll",
+  ComputeLayout = "ComputeLayout",
+  DiscardDiagram = "DiscardDiagram",
+  DiscardStepSequence = "DiscardStepSequence",
+  Init = "Init",
+  LabelMeasurements = "LabelMeasurements",
+  Resample = "Resample",
+}
+
+// Request Data
+
+export type CompileRequestData = {
+  tag: MessageTags.Compile;
+  domain: string;
+  style: string;
+  substance: string;
+  variation: string;
+};
+
+export type PollRequestData = {
+  tag: MessageTags.Poll;
+  diagramId: number;
+};
+
+export type ComputeLayoutRequestData = {
+  tag: MessageTags.ComputeLayout;
+  diagramId: DiagramID;
+  historyLoc: HistoryLoc;
+};
+
+export type ResampleRequestData = {
+  tag: MessageTags.Resample;
+  diagramId: DiagramID;
+  variation: string;
+};
+
+// Notification Data
+
+export type InitData = {
+  tag: MessageTags.Init;
+};
+
+export type LabelMeasurementData = {
+  tag: MessageTags.LabelMeasurements;
+  diagramId: DiagramID;
+  labelMeasurements: LabelMeasurements;
+};
+
+export type DiscardDiagramData = {
+  tag: MessageTags.DiscardDiagram;
+  diagramId: DiagramID;
+};
+
+// Results
+
+export type TaggedOk<V, T> = {
+  tag: T;
+  variant: "Ok";
+  value: V;
+};
+
+export type TaggedErr<E, T> = {
+  tag: T;
+  variant: "Err";
+  error: E;
+};
+
+/**
+ * A Result type, but with a tag. Supertype of `MessageResult`.
+ */
+export type TaggedResult<V, E, T> = TaggedOk<V, T> | TaggedErr<E, T>;
+
+export type CompileResult = TaggedResult<
+  CompileInfo,
+  PenroseError,
+  MessageTags.Compile
+>;
+
+export type PollResult = TaggedResult<
+  HistoryInfo,
+  InvalidDiagramIDError,
+  MessageTags.Poll
+>;
+
+export type ComputeLayoutResult = TaggedResult<
+  LayoutState,
+  InvalidDiagramIDError | InvalidHistoryLocError | PenroseError,
+  MessageTags.ComputeLayout
+>;
+
+export type ResampleResult = TaggedResult<
+  StepSequenceID,
+  InvalidDiagramIDError,
+  MessageTags.Resample
+>;
+
+// Errors
+
+export type InvalidDiagramIDError = {
+  tag: "InvalidDiagramIDError";
+  id: DiagramID;
+  validIds: Set<DiagramID>;
+};
+
+export type InvalidHistoryLocError = {
+  tag: "InvalidHistoryLocError";
+  historyLoc: HistoryLoc;
+  historyInfo: HistoryInfo;
+};
+export type OptimizationError = {
+  tag: "OptimizationError";
+  error: unknown;
+};
+
+export type OptimizerError =
+  | OptimizationError
+  | InvalidDiagramIDError
+  | InvalidHistoryLocError;
+
+// Utils
+
+class NumericIDGenerator {
+  private nextId = 0;
+
+  next = () => {
+    return this.nextId++;
+  };
+}
+
+// We reexport as different generators for each ID type, in case we decide
+// not to use numbers in the future
+export {
```

Comment:
No need to do this now, but a tool that might be useful in the future, in case we want them to stay numbers but also be different types statically: https://egghead.io/blog/using-branded-types-in-typescript

Reply:
Oh, sweet! That's perfect. Thanks!